unittest/unittest.py

```diff
@@ -176,15 +179,30 @@ def __init__(self):
         self.failuresNum = 0
         self.skippedNum = 0
         self.testsRun = 0
+        self.exceptions = []
 
     def wasSuccessful(self):
         return self.errorsNum == 0 and self.failuresNum == 0
 
+if hasattr(sys, 'print_exception'):
```

Comment:
As you see around this file, doublesquotes are used. That's my default style for strings. So, I would ask you do update your changes (more cases elsewhere).

As for this chunk of code, well, I remember some people wanted to use unittest on devices. I guess, this can be seen as courtesy to them (instead of depending on traceback module right away). But note that feature you add, accumulating large objects in memory, goes against such usage in the first place. Nothing needs changing here, just food for thought.

Reply:
Point noted on string style, thanks. I'll clean up pre-existing inconsistencies while I'm at it.

With the print_exception stuff, yes I'm using unittest on both unix and stm32; unix for development and CI, stm32 for product proof / hardware in the loop testing. 
This allows us to run the exact same tests on both CI and on hardware via `pyboard.py`

I'm probably up against the limit of my knowledge of the vm & gc, would you mind highlighting the biggest costs of this implementation of a `print_exception` function / handle?  I do want to make my work as efficient as practical.
Other than the parsing and compiling costs of more code text, is it the lamba usage adding the extra layer of abstraction? Does having it at global scope make it worse? Am I missing something bigger?

That block I did copy from some other library, looking again it really does not need to be standalone and the functionality should just be moved into `def capture_exc(e)`.

Reply:
> I'll clean up pre-existing inconsistencies while I'm at it.

No, please ;-). Just use the style the majority of file uses. (And you now know what should be majority ;-) ).

As separate work/PRs, might make such cleanups of course, but there're so many ways to spend time on micropython-lib in more useful ways, so I'd discourage spend it on strings either.

P.S. There're tools which can do that automatically.

Reply:
As for other questions, I'm not sure where they come from. For some other module - sure, may makes sense to think about all those things (and great that you do). But for "unittest"? Nah. 

"But note that feature you add, accumulating large objects in memory" means just what it says:

You do:

> exceptions.append(...)

So, if each of 1000 tests throws exception, you'll have a list with 1000 exception tracebacks. That's fully ok for this kind of module from my PoV.

Reply:
Ah yes, I completely follow you now. Yes it's a risky pattern to follow, unbound length arrays growing like that.

Reply:
Not "risky", rather "desktoppy" and not "embeddy". As unittest isn't really intended for embedded usage, that's ok. But knowing that some people (like yourself) use it on embedded is helpful. (When needed, we can split of **u**unittest, though I'd hope that for embedded usage, someone would come up with purposedly designed module at all.)

Reply:
Even on a desktop it's often bad practice, but yeah it's usually a deal breaker on embedded. My background is in electronics and small embedded work, but I find myself getting very lazy with python which is a dangerous mentality to get into (even though our stm32 board has 32MB of sdram).

I've considered splitting modules myself for smaller embedded targeting version, though I'd want some significant differences between it and a full unix version to be worth maintaining two copies. 
I'm working on a pathlib port from the cpython impl in the background at the moment though and that's a definite candidate for two versions.... the initially mostly working copy is still a 42KB python file, that's not going on many embedded platforms as is for sure.